components/autocomplete/index.js

```diff
@@ -225,22 +226,85 @@ export class Autocomplete extends Component {
 		this.node = node;
 	}
 
-	insertCompletion( range, replacement ) {
-		const container = document.createElement( 'div' );
-		container.innerHTML = renderToString( replacement );
-		while ( container.firstChild ) {
-			const child = container.firstChild;
-			container.removeChild( child );
-			range.insertNode( child );
-			range.setStartAfter( child );
+	insertCompletion( range, replacement, completerName ) {
+		// Wrap completions so we can treat them as tokens with editing boundaries.
+		const tokenWrapper = document.createElement( 'span' );
+		tokenWrapper.innerHTML = renderToString( replacement );
```

Comment:
What if the replacement is an element? Should we maybe only wrap gin a span when it's a string?

Reply:
Thanks, @iseulde, I think that's a great suggestion. If a completer wants to insert custom HTML, it could, and we wouldn't interfere by wrapping it in a token span. I am out of time today but plan to address this in the morning.

Reply:
Actually, I'm not sure what to do here because, even if we don't wrap with a span, we probably want editing boundaries for any autocompletion. What do you think?

Should we add a class to the element if a completer returns one? What if a completer returns a fragment? This may have been the line of thinking I followed to wrap everything in a span, though I understand not all elements are valid children of a span.

Reply:
After sleeping on this, my thinking is:
* String completions should be wrapped as editable tokens. I believe this is the simple autocompletion experience we often expect.
* Element completions should be marked as read-only since it is arbitrary content and may be awkward to edit properly.

What do you think, @mtias?

Reply:
It sounds like there could be two types of completers, but given lack of examples of the second kind, we could optimize for the cases we have now.

Reply:
Here is an example of the second kind where a completer is used to insert an acronym:
https://gist.github.com/brandonpayton/01b75c55ce4d329d534b1d3a68dc728f

![second-completer-type-example](https://user-images.githubusercontent.com/530877/41986081-f3685830-79e9-11e8-833e-009499a51aed.gif)

In this case, the completion does not include the trigger prefix, and its `<abbr>` element has a `title` attribute that cannot be edited via the RichText component. My take is that we should not wrap these in the same way but should mark them as readonly so the editor can remove but not edit them. Does that example help?

Reply:
Sounds good to me.